File: backend/app/schemas/cartao.py
"""Schemas de Cartao: payloads de entrada e saída de `app/api/routes/cartao.py`.

`limite_disponivel` existe SÓ em `CartaoRead` - não é uma coluna do model
`Cartao` (mesmo princípio de `ContaRead.saldo_atual`: valor sempre
calculado, nunca armazenado). `CartaoService` calcula esse valor e o anexa
como atributo transiente ao objeto `Cartao` antes de devolvê-lo ao Router.
"""
from decimal import Decimal

from pydantic import BaseModel, Field

from app.models.enums import Bandeira
from app.schemas.base import OrmBaseModel

_PADRAO_ULTIMOS_QUATRO_DIGITOS = r"^\d{4}$"


class CartaoCreate(BaseModel):
    nome: str = Field(min_length=1, max_length=120)
    conta_pagamento_id: int
    instituicao: str = Field(min_length=1, max_length=120)
    bandeira: Bandeira
    ultimos_quatro_digitos: str = Field(pattern=_PADRAO_ULTIMOS_QUATRO_DIGITOS)
    limite: Decimal = Field(ge=0)
    dia_fechamento: int = Field(ge=1, le=31)
    dia_vencimento: int = Field(ge=1, le=31)
    # "Estado Inicial do Cartao" - quanto do limite ja estava em uso quando
    # o cartao foi cadastrado, informado direto na criacao (sem Fatura/
    # Transacao por tras). Opcional - default 0 (cartao novo, sem uso
    # anterior), ver docstring de `Cartao.saldo_inicial_utilizado`.
    saldo_inicial_utilizado: Decimal = Field(default=Decimal("0"), ge=0)


class CartaoUpdate(BaseModel):
    """Todos os campos são opcionais - só o que for enviado é alterado
    (semântica de PATCH). `usuario_id` nunca aparece aqui: a posse de um
    cartão não é algo que se atualiza via payload."""

    nome: str | None = Field(default=None, min_length=1, max_length=120)
    conta_pagamento_id: int | None = None
    instituicao: str | None = Field(default=None, min_length=1, max_length=120)
    bandeira: Bandeira | None = None
    ultimos_quatro_digitos: str | None = Field(default=None, pattern=_PADRAO_ULTIMOS_QUATRO_DIGITOS)
    limite: Decimal | None = Field(default=None, ge=0)
    dia_fechamento: int | None = Field(default=None, ge=1, le=31)
    dia_vencimento: int | None = Field(default=None, ge=1, le=31)
    ativo: bool | None = None
    saldo_inicial_utilizado: Decimal | None = Field(default=None, ge=0)


class CartaoRead(OrmBaseModel):
    id: int
    nome: str
    conta_pagamento_id: int
    instituicao: str
    bandeira: Bandeira
    ultimos_quatro_digitos: str
    limite: Decimal
    limite_disponivel: Decimal
    dia_fechamento: int
    dia_vencimento: int
    ativo: bool
    saldo_inicial_utilizado: Decimal

File: frontend/src/types/cartao.ts
/**
 * Espelha 1:1 `app/schemas/cartao.py` — conferido por leitura direta do
 * backend (`CartaoCreate`/`CartaoUpdate`/`CartaoRead`). Mesmo tratamento de
 * `types/conta.ts`: todo valor monetário é `string` (nunca `number`),
 * `limite_disponivel` existe só em `CartaoRead` (calculado pelo
 * `CartaoService` a cada leitura, nunca uma coluna — mesmo princípio de
 * `ContaRead.saldo_atual`, **pode ficar negativo de propósito**, nunca
 * "clampado" em zero).
 */
import type { Bandeira } from "./enums";

export type { Bandeira };

export interface CartaoRead {
  id: number;
  nome: string;
  conta_pagamento_id: number;
  instituicao: string;
  bandeira: Bandeira;
  ultimos_quatro_digitos: string;
  limite: string;
  limite_disponivel: string;
  dia_fechamento: number;
  dia_vencimento: number;
  ativo: boolean;
  /** "Estado Inicial do Cartão" — quanto do limite já estava em uso quando
   * o cartão foi cadastrado, declarado direto (sem Fatura/Transacao por
   * trás). Consome `limite_disponivel` permanentemente até o usuário
   * editar/zerar (Sprint de Refinamento Premium, 2026-07). */
  saldo_inicial_utilizado: string;
}

export interface CartaoCreate {
  nome: string;
  conta_pagamento_id: number;
  instituicao: string;
  bandeira: Bandeira;
  ultimos_quatro_digitos: string;
  limite: string;
  dia_fechamento: number;
  dia_vencimento: number;
  saldo_inicial_utilizado?: string;
}

/** Semântica de PATCH — todo campo omitido permanece intocado. Reativar um
 * cartão desativado é `PATCH { ativo: true }`, mesmo endpoint de edição
 * (mesma decisão de Conta/Categoria/Tag). */
export interface CartaoUpdate {
  nome?: string;
  conta_pagamento_id?: number;
  instituicao?: string;
  bandeira?: Bandeira;
  ultimos_quatro_digitos?: string;
  limite?: string;
  dia_fechamento?: number;
  dia_vencimento?: number;
  ativo?: boolean;
  saldo_inicial_utilizado?: string;
}
